### DISCUSSION TITLE
ViSP bottle needs to be updated

### DISCUSSION THREAD

**Discussion by fspindle** [Homebrew/discussions#5787] [Writing Formulae/Casks]
### Output of `brew config`

```shell
% brew config
HOMEBREW_VERSION: 4.4.9
ORIGIN: https://github.com/Homebrew/brew
HEAD: 2e780004c92f16a96036d3d43569c2924c4e54f5
Last commit: 3 days ago
Branch: stable
Core tap JSON: 04 Dec 08:34 UTC
Core cask tap JSON: 04 Dec 08:34 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_DISPLAY: /private/tmp/com.apple.launchd.hIAhMy71s4/org.xquartz:0
HOMEBREW_MAKE_JOBS: 10
Homebrew Ruby: 3.3.6 => /opt/homebrew/Library/Homebrew/vendor/portable-ruby/3.3.6/bin/ruby
CPU: 10-core 64-bit arm_firestorm_icestorm
Clang: 16.0.0 build 1600
Git: 2.47.1 => /opt/homebrew/bin/git
Curl: 8.7.1 => /usr/bin/curl
macOS: 14.7.1-arm64
CLT: 16.1.0.0.1.1729049160
Xcode: 16.1
Rosetta 2: false
```

### Output of `brew doctor`

```shell
% brew doctor
Your system is ready to brew.
```

### Description of issue

[ViSP](https://github.com/Homebrew/homebrew-core/blob/master/Formula/v/visp.rb) bottle is no more compatible with a recent update of [OpenCV](https://github.com/Homebrew/homebrew-core/blob/master/Formula/o/opencv.rb) bottle (OpenCV is a dependency of ViSP).
See https://github.com/lagadic/visp/issues/1515

```
$ brew install opencv
```
will bring OpenCV installation in `/opt/homebrew/Cellar/opencv/4.10.0_14/` folder while ViSP installed with
```
$ brew install visp
```
will contain in `/opt/homebrew/Cellar/visp/3.6.0_6/lib/cmake/visp/VISPConfig.cmake` a reference to `/opt/homebrew/Cellar/opencv/4.10.0/include/opencv4` instead of `/opt/homebrew/Cellar/opencv/4.10.0_14/include/opencv4`

Can some one (@cho-m, @issyl0, @chenrui333, @branchvincent, @fxcoudert ...) update visp bottle ?

**Comment by p-linnane** [comment]
Thanks for the report. I've opened https://github.com/Homebrew/homebrew-core/pull/200220 to rebuild the bottle.

**Comment by sulibo** [comment]
OpenCV was updated two days ago, now it is 4.10.0_15, see [update 4.10.0_15 bottle](https://github.com/Homebrew/homebrew-core/commit/f63f5baa0e1fad7c4834703e25d4c1f64f4a4787). Is it possible to update the visp bottle accordingly? Thanks!

**Reply by p-linnane** [reply]
If `visp` needs to rebuilt every time `opencv` is, then we need to update the `visp` test to fail so we can catch this.

**Comment by alebcay** [comment]
Can you share more about what the cmake file looks like for which you are seeing the issues? Tentatively, I have tried something like the following and it seems to be working fine (maybe I'm not testing a part that relies on OpenCV):

```cmake
cmake_minimum_required(VERSION 3.10 FATAL_ERROR)
project(visp-check LANGUAGES CXX)
find_package(VISP REQUIRED visp_core)
include_directories(${VISP_INCLUDE_DIRS})
add_executable(visp-check test.cpp)
target_link_libraries(visp-check ${VISP_LIBRARIES})
```

**Reply by sulibo** [reply]
Sure, here is the cmake file I used: 

```cmake
cmake_minimum_required(VERSION 3.10)
project(tutorial-image)
find_package(VISP REQUIRED)
include_directories(${VISP_INCLUDE_DIRS})
add_executable(tutorial-viewer tutorial-viewer.cpp)
target_link_libraries(tutorial-viewer ${VISP_LIBRARIES})
```
More details can be found at https://github.com/lagadic/visp/issues/1515.

**Reply by cho-m** [reply]
> More details can be found at [lagadic/visp#1515](https://github.com/lagadic/visp/issues/1515).

There aren't any details on running with the fixed bottle (https://github.com/Homebrew/homebrew-core/pull/200267).

---

Can you make sure you are on the latest revision of `visp` (`brew info visp` should show **/usr/local/Cellar/visp/3.6.0_8**) and provide output from running CMake

**Comment by fspindle** [comment]
It sounds that the last ViSP bottle is now synchronised with OpenCV.

As mentioned by @p-linnane bellow is a test that could be added to check compat between ViSP and OpenCV

`CMakeLists.txt` file content
```
cmake_minimum_required(VERSION 3.10)

project(test-brew-opencv)

find_package(VISP)
add_executable(test-opencv test-opencv.cpp)
target_include_directories(test-opencv PRIVATE ${VISP_INCLUDE_DIRS})
target_link_libraries(test-opencv ${VISP_LIBRARIES})
```

and `test-opencv.cpp` file content
```
#include <visp3/core/vpImage.h>
#include <visp3/core/vpImageConvert.h>
#include <visp3/core/vpConvert.h>

int main()
{
  cv::Mat mat(3, 3, CV_8UC1, 10);
  vpImage<unsigned char> I;
  vpImageConvert::convert(mat, I);
  for (unsigned int i = 0; i < mat.rows * mat.cols; ++i) {
    if (I.bitmap[i] != 10) {
      std::cout << "Error in image conversion from opencv to visp" << std::endl;
      return EXIT_FAILURE;
    }
  }

  vpImagePoint ip(10, 20);
  cv::Point2d cv_point2d;
  vpConvert::convertToOpenCV(ip, cv_point2d);

  if ((ip.get_u() != cv_point2d.x) || (ip.get_v() != cv_point2d.y)) {
    std::cout << "Error in point2d conversion from visp to opencv" << std::endl;
    return EXIT_FAILURE;
  }

  std::cout << "Test succeed" << std::endl;
  return EXIT_SUCCESS;
}
```
